### PR TITLE
API Auth Ingest/Consumption

### PR DESCRIPTION
**Pull request by georgevanderson**
Set the hashed values as environment variables to turn on auth
MOOSE_INGEST_API_KEY
MOOSE_CONSUMPTION_API_KEY

### CODE REVIEW

File: apps/framework-cli/src/cli/local_webserver.rs [hunk 1]
```diff
@@ -530,6 +574,17 @@ async fn ingest_route(
...
 
+    let auth_header = req.headers().get(hyper::header::AUTHORIZATION);
+
+    if !check_authorization(auth_header, "MOOSE_INGEST_API_KEY").await {
+        return Response::builder()
+            .status(StatusCode::UNAUTHORIZED)
+            .body(Full::new(Bytes::from(
+                "Unauthorized: Invalid or missing token",
+            )))
+            .map_err(Into::into);
```
Comment (phiSgr) [inline apps/framework-cli/src/cli/local_webserver.rs]: looks like we don't need `.map_err(Into::into);`

File: apps/framework-cli/src/cli/local_webserver.rs [hunk 2]
```diff
@@ -515,6 +530,35 @@ async fn handle_json_array_body(
...
+        },
+        Err(_) => {
+            // If API_KEY not SET
+            true
+        }
+    }
+}
+
+async fn check_authorization(auth_header: Option<&HeaderValue>, env_var: &str) -> bool {
+    let bearer_token = match auth_header {
```
Comment (phiSgr) [inline apps/framework-cli/src/cli/local_webserver.rs]: maybe use `and_then`?

https://doc.rust-lang.org/std/option/enum.Option.html#method.and_then

File: apps/framework-cli/src/cli/local_webserver.rs [hunk 3]
```diff
@@ -515,6 +530,35 @@ async fn handle_json_array_body(
     success_response(url)
 }
 
+async fn validate_token(token: Option<&str>, env_var: &str) -> bool {
+    match env::var(env_var) {
```
Comment (phiSgr) [inline apps/framework-cli/src/cli/local_webserver.rs]: we can probably use a `lazy_static` to cache the value

File: apps/framework-cli/src/cli/local_webserver.rs [hunk 4]
```diff
@@ -515,6 +530,35 @@ async fn handle_json_array_body(
...
+        Ok(ingest_api_key) => match token {
```
Comment (phiSgr) [inline apps/framework-cli/src/cli/local_webserver.rs]: `token.is_some_and`

File: apps/framework-cli/src/cli/local_webserver.rs [hunk 5]
```diff
@@ -103,7 +107,18 @@ async fn create_client(
...
+    // Extract the Authorization header and check the bearer token
+    let auth_header = req.headers().get(hyper::header::AUTHORIZATION);
+
+    if !check_authorization(auth_header, "MOOSE_CONSUMPTION_API_KEY").await {
+        return Response::builder()
+            .status(StatusCode::UNAUTHORIZED)
+            .body(Full::new(Bytes::from(
+                "Unauthorized: Invalid or missing token",
+            )))
+            .map_err(Into::into);
```
Comment (phiSgr) [inline apps/framework-cli/src/cli/local_webserver.rs]: i prefer `return Ok(...?);`

**Review by phiSgr — COMMENTED** [review]
just some code style stuff

File: apps/framework-cli/src/cli/local_webserver.rs [hunk 6]
```diff
@@ -515,6 +530,35 @@ async fn handle_json_array_body(
     success_response(url)
 }
 
+async fn validate_token(token: Option<&str>, env_var: &str) -> bool {
+    match env::var(env_var) {
+        Ok(ingest_api_key) => match token {
+            Some(token) => hex_digest(Algorithm::SHA256, token.as_bytes()) == ingest_api_key,
```
Comment (callicles) [inline apps/framework-cli/src/cli/local_webserver.rs]: We should not be using SHA-256 for this.

NIST recommends PBKDF2 - let's follow that recommendation with a constant time check on the output.

Ref: https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2

https://github.com/RustCrypto/password-hashes/tree/master/pbkdf2

**Review by callicles — CHANGES_REQUESTED** [review]
Let's change the algorithm used for hashing

File: apps/framework-cli/src/cli/routines/auth.rs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+    let key1_hex = hex::encode(key1);
+
+    // compare byte to byte to avoid timing attacks
+    let token_hash = hex::decode(private_pass).unwrap();
+    let key1_hash = hex::decode(key1_hex).unwrap();
+
+    for i in 0..token_hash.len() {
+        if token_hash[i] != key1_hash[i] {
+            return false;
```
Comment (callicles) [inline apps/framework-cli/src/cli/routines/auth.rs]: If I am reading this correcly - you are doing an early return here - so not comparing byte. 

Doing constant time comparison is really hard - I would use a third-party to do that.

After some research I would use this: https://github.com/cesarb/constant_time_eq/tree/master

**Review by callicles — CHANGES_REQUESTED** [review]
Sorry - last change and then we will be good to go 😄 Security is hard

**Review by callicles — APPROVED** [review]
Nice work!

**Pull request merged**